Hosting Command Palette: Add README with basic instructions (#84591)
Automattic/wp-calypso: client/components/command-pallette/README.md

# Command Palette

The Command Palette is a React component used to display a text input used to search and filter a list of commands. It can be accessed in the Calypso using the shortcut `cmd+k`.

The command palette is inspired by [Core](https://github.com/WordPress/gutenberg/blob/trunk/packages/commands/README.md) implementation used in the Gutenberg editor.

## Add a command

> In order to add a new command to the palette, you can add them in `useCommandsArrayWpcom` or create a new hook and include it to `useCommandPallette` hooks.

Reviewer: ```suggestion
In order to add a new command to the palette, you can add it in `useCommandsArrayWpcom` or create a new hook and include it to `useCommandPallette` hooks.
```

Author: Changed! https://github.com/Automattic/wp-calypso/pull/84591/commits/b274e4d2a752bbbfe2e05481b9563281b45c45f6

## Command properties

Every command has a set of properties that can be used to customize its behavior. The following properties are available:

> - `name`: Type `string`, used as identifier for the command. It's ex

Reviewer: Something got cut here?

Author: Completed 👌  https://github.com/Automattic/wp-calypso/pull/84591/commits/7c545edafef9a30aa2f90984cc7dc53059c533ca

- `label`: Type `string`, used as the label for the command. Visible in the command palette.
- `subLabel`?: Type `string`, used as a second line for the command. Visible in the command palette.
- `searchLabel`?: Type `string`, used to match the user search. It's not visible for the user. We recommend to use `_x()` to add context to the translators.

> - `callback`: Type `function`, used to execute the command when the user clicks on it or press enter on that command.

Reviewer: ```suggestion
- `callback`: Type `function`, used to execute the command when the user clicks on it or presses enter on that command.
```

Author: Changed https://github.com/Automattic/wp-calypso/pull/84591/commits/09841b779732db8a0f5214444513d9a07d139c5a

> - `context`?: Type `string[]`, a list of paths used to display the command by default.

Reviewer: I feel we could clarify it more. I'm not sure if it's the best wording, though: 

```suggestion
- `context`?: Type `string[]`, a list of URL paths for which the command should be displayed.
```

Author: Changed https://github.com/Automattic/wp-calypso/pull/84591/commits/44a1933d989bb36da1e6b51b2edca86642d7e73b

- `icon`?: Type `JSX.Element`, used to display an icon for the command. It's visible in the command palette.
- `image`?: Type `JSX.Element`, used to display an image for the command, instead of an icon. It's visible in the command palette.
- `siteFunctions`?: Type `SiteFunctions`, used for nested commands that need to execute a function when a site is selected in the command palette as a second step.

> ## Usage

Reviewer: What do you think about showing how to define example command, too?

Author: I added a command example:

https://github.com/Automattic/wp-calypso/pull/84591/commits/6e4db1601bac2cfa95076542ce33600f14ee13dc

Reviewer: Thanks! Do you think we should keep this example simple, or could we use it to show how to use other fields e.g. `searchLabel`?

```
searchLabel: _x( 'add / create a new site', 'used as a keywords when user searches for commands' ),
```

Reviewer 2: Seems like example should be first up, and embedding bits aren't something folks need to worry about so it can be lower.

# Command Palette

The Command Palette is a React component used to display a text input used to search and filter a list of commands. It can be accessed in the Calypso using the shortcut `cmd+k`.

> The command palette is inspired by [Core](https://github.com/WordPress/gutenberg/blob/trunk/packages/commands/README.md) implementation used in the Gutenberg editor.

Reviewer 3: Can we describe how the WordPress.com version is similar to and different from the core version?

Author: Added here https://github.com/Automattic/wp-calypso/pull/84591/commits/1986403aaf6185b9095f8b2eb45381acf77447fe and https://github.com/Automattic/wp-calypso/pull/84591/commits/80de364a3dad39f98296526c5f5bafa626c2f4ce

## Add a command

> In order to add a new command to the palette, you can add it in `useCommandsArrayWpcom` or create a new hook and include it to `useCommandPallette` hooks.

Reviewer 3: Couple of questions here:

1. Can we provide guidance on "best practices" for adding a command? e.g. What makes a good command vs. not, what language should be used to describe the command, etc.?
2. Should we have all WordPress.com commands registered via a central mechanism, so it's easy to keep tabs on what commands are added? I think that developers adding commands shouldn't have to worry about mounting the `<WpcomCommandPalette />` component.

Author: > Can we provide guidance on "best practices" for adding a command? e.g. What makes a good command vs. not, what language should be used to describe the command, etc.?

Added here https://github.com/Automattic/wp-calypso/tree/add/command-palette-add-readme/client/components/command-pallette#best-practices-for-defining-a-command

> Should we have all WordPress.com commands registered via a central mechanism, so it's easy to keep tabs on what commands are added? I think that developers adding commands shouldn't have to worry about mounting the <WpcomCommandPalette /> component.

I was more explicit that we already are mounting `WpcomCommandPalette` in "all" Claypso pages for WPcom. So developers do not need to mount it again, just add a command into the array.

- The Calypso version supports multisite commands with the sites as second step commands.
- Users can go back to root commands with the `esc` and `backspace` keys, or the back button.
- We change the search placeholder to inform the user about the current command.
- Besides of icons, we also support images for the commands.
- We support two lines of text for the commands.

## Add a command

In order to add a new command to the palette, you can include the command object in `useCommandsArrayWpcom` or create a new hook and include it to `useCommandPallette` hooks.

## Command properties

Every command has a set of properties that can be used to customize its behavior. The following properties are available:

- `name`: Type `string`, used as identifier for the command. It's expected to be unique and `camelCase`. It's not visible for the user.
- `label`: Type `string`, used as the label for the command. Visible in the command palette.
- `subLabel`?: Type `string`, used as a second line for the command. Visible in the command palette.

> - `searchLabel`?: Type `string`, used to match the user search. It's not visible for the user. We recommend to use `_x()` to add context to the translators.

Reviewer 2: You could add example of context string, so that all commands could use the same consistently.

Author: I added the example https://github.com/Automattic/wp-calypso/pull/84591/commits/903b3fbb5ea153fff2014145d965c3b8c5b5ef7e . There is also a full command example a bit below.

- `callback`: Type `function`, used to execute the command when the user clicks on it or presses enter on that command.
- `context`?: Type `string[]`, a list of URL paths for which the command will have priority when displayed.
- `icon`?: Type `JSX.Element`, used to display an icon for the command. It's visible in the command palette.
- `image`?: Type `JSX.Element`, used to display an image for the command, instead of an icon. It's visible in the command palette.

> - `siteFunctions`?: Type `object`, used for nested commands that need to execute a function when a site is selected in the command palette as a second step.

Reviewer 2: Could clarify what happens if there's already active site, i.e. most pages outside `/sites`, `/read` and `/me`.

Author: I added the current behaviour here https://github.com/Automattic/wp-calypso/pull/84591/commits/4c9db3b0b165842c81d89950ba3dbf13c14e4d49

After trying the palette globally I see we can iterate on the behaviour, and for selected sites it could navigate directly between pages. My other idea for when a site is selected, we can display the current site as first option always.

Reviewer 2: Additional site switcher upon commands feels like a lot of friction for what's supposed to be quick commands, if the site is already "selected". There's the quick-site switcher in the sidebar, after all.

Reviewer 3: I agree. In the single site context, the command should apply to the current site and you shouldn't see the site switcher.

## Best practices for defining a command

Please consider carefully before adding a new command to the palette. We aim to avoid incorporating commands that are infrequently used and unlikely to be removed in the future.

- Command labels should be short and descriptive and use sentence case. For example, "Add new site" instead of "Add New Site".
- Commands should start with a verb. For example, "Add new site" instead of "New site".

> - Picking the correct verb is important. `Open` is the more generic verb, if the page has a list of elements, we can use `View`, if the user is expected to take an action we can use `Manage` or `Change`. We prefer specific verbs over generic ones.

Reviewer 2: Another aspect is to stick with terminology used elsewhere in the interace, and use aliases for synonyms.

Examples:
- "Import a site" because we have Tools → Import, and alias with search label "migrate a site".
- "Add new post" because "add" is used in WP instead of "write", which still makes a good synonym as a search label.

Author: Done here https://github.com/Automattic/wp-calypso/pull/84591/commits/13158f7b2c189c0993e70e8ec224cd08cb89634c

# Command Palette

> The Command Palette is a React component used to display a text input used to search and filter a list of commands. It can be accessed in the Calypso using the shortcut `cmd+k`.

Reviewer 3: ```suggestion
The Command Palette is a UI element that displays a searchable list of commands. It can be accessed in the Calypso using the shortcut `cmd+k`.
```

Author: Done ! https://github.com/Automattic/wp-calypso/pull/84591/commits/711b1777e3d6a379192a949cd5f5163bb2320c15

> The command palette is inspired by [Core](https://github.com/WordPress/gutenberg/blob/trunk/packages/commands/README.md) implementation used in the Gutenberg editor. We use the same CSS and similar markup, the commands share the same basic props that we extended to support the following differences:

Reviewer 3: ```suggestion
This command palette is inspired by [Core's implementation](https://github.com/WordPress/gutenberg/blob/trunk/packages/commands/README.md) used in the Gutenberg editor. We use the same CSS and similar markup, the commands share the same basic props that we extended to support the following differences:
```

Author: Done ! https://github.com/Automattic/wp-calypso/pull/84591/commits/70845f2a43ae7a33e432373f01f42dece90da1cf

- The Calypso version supports multisite commands with the sites as second step commands.
- Users can go back to root commands with the `esc` and `backspace` keys, or the back button.
- We change the search placeholder to inform the user about the current command.

> - Besides of icons, we also support images for the commands.

Reviewer 3: ```suggestion
- In addition to icons, we also support images for the commands.
```

Author: Done! https://github.com/Automattic/wp-calypso/pull/84591/commits/f380f7ce516ccaf0e0cccba8a88e92dd513cc4ac

- We support two lines of text for the commands.

> 

Reviewer 3: Can we include some mention of https://github.com/WordPress/gutenberg/issues/55514 and how we'd eventually like to reconcile the differences?

Author: Done! https://github.com/Automattic/wp-calypso/pull/84591/commits/6ea1b16884a6a5c6ef4076d8b9f96d54e18e5dd1

## Add a command

In order to add a new command to the palette, you can include the command object in `useCommandsArrayWpcom` or create a new hook and include it to `useCommandPallette` hooks.

> ## Command properties

Reviewer 3: ```suggestion
### Command properties
```

Author: Done https://github.com/Automattic/wp-calypso/pull/84591/commits/541437673066c2496ebcc5db6d5ed2d310c1fa52

- `label`: Type `string`, used as the label for the command. Visible in the command palette.
- `subLabel`?: Type `string`, used as a second line for the command. Visible in the command palette.
- `searchLabel`?: Type `string`, used to match the user search. It's not visible for the user. We recommend to use `_x()` to add context to the translators.
- `callback`: Type `function`, used to execute the command when the user clicks on it or presses enter on that command.
- `context`?: Type `string[]`, a list of URL paths for which the command will have priority when displayed. Example: `['/sites', '/manage/domains']`.
- `icon`?: Type `JSX.Element`, used to display an icon for the command. It's visible in the command palette.
- `image`?: Type `JSX.Element`, used to display an image for the command, instead of an icon. It's visible in the command palette.
- `siteFunctions`?: Type `object`, used for nested commands that need to execute a function when a site is selected in the command palette as a second step. The behaviour for nested commands is the same independently of the page we are.

> ## Best practices for defining a command

Reviewer 3: ```suggestion
### Best practices for defining a command
```

Author: Done! https://github.com/Automattic/wp-calypso/pull/84591/commits/541437673066c2496ebcc5db6d5ed2d310c1fa52

- Command labels should be short and descriptive and use sentence case. For example, "Add new site" instead of "Add New Site".
- Commands should start with a verb. For example, "Add new site" instead of "New site".
- Picking the correct verb is important. `Open` is the more generic verb, if the page has a list of elements, we can use `View`, if the user is expected to take an action we can use `Manage` or `Change`. We prefer specific verbs over generic ones.
- We prefer to use [Gutenberg icons](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library) for root commands.
- It's important to keep the same terminology and icons used elsewhere in the interace or in core.
- We can introduce aliases, and keywords using `searchLabel`. For example. "Import a site" would be an alias of "Migrate a site" because the navigation it's under Tools → Import.
- Most commands navigate to other pages. We can also navigate to a specific section/card inside a page. For example, "Manage cache settings" in addition to "Manage hosting configuration".

> - Commands that open modals are also allowed. Just be aware that the command palette is visible in most Calypso pages.

Reviewer 3: > Just be aware that the command palette is visible in most Calypso pages.

What are the implications of this? That the modal will need to be loaded on those pages too?

Author: Yes, that's what would happen by default.
I added a comment to consider loading the modal async https://github.com/Automattic/wp-calypso/pull/84591/commits/2ac72b883479fdeca1ca9ce6a0c72846868fa2c9
